### DISCUSSION TITLE
EasyMC / MCLeaks Support

### DISCUSSION THREAD

**Discussion by LucasLauDev** [PrismarineJS/mineflayer#2575] [General]
Well, well, well. I know many of you trying to make mineflayer support for mcleaks or easymc auth, and It's not just only me who made it successfully. Some of them choose to use it themselves and I choose to share. As I am not very rich having many accounts and the server is allowed only one username per server, so I have look into their API, and nothing I get, I have try for many times, and give up around 5 times. After that I have decompiler their mods and app, then I noticed something about their auth server. Then, I fork this repo, and made changes in auth server part. Finally, I got it works. 
Github: https://github.com/Happy-and-Smile/minecraft-player

## How to install
1. You have to uninstall the original mineflayer if you have installed it
2. Install the package through `npm i github:Happy-and-Smile/minecraft-player`

## Example Code For EasyMC
```js
const mineflayer = require('mineflayer')

const bot = mineflayer.createBot({
   host: 'localhost',           // minecraft server ip
  // port: 25565,               // only set if you need a port that isn't 25565
   version: "1.18.2",           // Minecraft Version
   auth: 'easymc',              // easymc or mcleaks or mojang or microsoft
   alt: ' ',                    // alt token
   sessionPath: 'session.json'  // session file suggested to be JSON type file
})

bot.on('chat', (username, message) => {
  if (username === bot.username) return
  bot.chat(message)
})

// Log errors and kick reasons:
bot.on('kicked', console.log)
bot.on('error', console.log)

```

## How To Automatic Renew Alt Token
You have to paid for bypass CloudFlare/Recaptcha, and look into the way to renew it.
So, the only one way is to save the session into a file, then you don't have to renew it everytime when you start the bot, but session will expire too. 

## Important Things
1. The repo is the same as mineflayer because it is a mineflayer fork (except adding easymc and mcleaks auth)
2. In case if there is any copyright, LICENSE or something else issue, I WILL NOT BE RESPONSIBLE, because I can't change the process without touching the original file and it will not be public to npmjs.com.
3. The repo will always update upstreams
4. The easymc or mcleaks servers sometimes will down

Sry for bad grammar because my englist not so good and bad coding skills because I am a "self learner" for JS.
If got any problem just let me know
Have A Nice Day.

**Comment by rom1504** [comment]
Hi QiuChen-Developer,

we don't really want to officially support mcleaks because it's in a gray area

however, I'd much prefer having a clean plugin than asking people to install an alternative version of mineflayer which may be outdated

do you want to create a mineflayer-mcleaks plugins ?

see https://github.com/PrismarineJS/node-minecraft-protocol/issues/736 and https://github.com/u9g/mineflayer-altlogin

**Reply by LucasLauDev** [reply]
I know that, but is it possible to load plugin before auth/login to the server?

**Reply by rom1504** [reply]
It should be possible in node-minecraft-protocol with the beforeLogin option
If it's not, making it possible is a good idea too

I think @u9g had a specific implementation idea

**Reply by u9g** [reply]
> I know that, but is it possible to load plugin before auth/login to the server?

My repo that rom linked isn't a plugin, it does the createbot for you and returns the bot

**Reply by mgi24** [reply]
Hi! im sorry if you don't support this plugin anymore, but i got problem while trying this. i got this error message.
>
options.session = options.sessionFile.find(x => x.alt === options.alt)
                                          ^
TypeError: Cannot read properties of undefined (reading 'find')

any fix or doc i can read? thanks!
